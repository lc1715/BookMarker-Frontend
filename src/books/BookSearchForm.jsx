import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
//Material UI
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import FormControl from '@mui/material/FormControl';
import OutlinedInput from '@mui/material/OutlinedInput';
import InputAdornment from '@mui/material/InputAdornment';
import IconButton from '@mui/material/IconButton';
import SearchIcon from '@mui/icons-material/Search';


function BookSearchForm() {
    const initialState = {
        term: '',
        title: '',
        author: '',
        isbn: '',
        keyword: ''
    }
    const [formData, setFormData] = useState(initialState);
    const [openAdvancedSearch, setOpenAdvancedSearch] = useState(false);
    const navigate = useNavigate();

    // Create advanced search term
    function advancedSearchTerm() {
        let advancedTerm = '';
        advancedTerm += formData.title ? `intitle:${formData.title}+` : '';
        advancedTerm += formData.author ? `inauthor:${formData.author}+` : '';
        advancedTerm += formData.isbn ? `inisbn:${formData.isbn}+` : '';
        advancedTerm += formData.keyword ? `${formData.keyword}` : '';

        if (advancedTerm[advancedTerm.length - 1] === '+') {
            advancedTerm = advancedTerm.slice(0, -1);
        }

        return advancedTerm;
    }

    // On form submit, check that search terms exist and navigate to BookList component
    async function handleSubmit(evt) {
        evt.preventDefault();
        let advancedTerm = advancedSearchTerm();

        if (formData.term === '' && advancedTerm === '') {
            return;
        }

        if (formData.term) {
            navigate(`/books/search/${formData.term}`);
        } else {
            navigate(`/books/search/${advancedTerm}`);
        }

        setFormData(initialState);
    };

    // Handles form data changing
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData((data) => ({ ...data, [name]: value }));
    }

    if (openAdvancedSearch) {
        return (
            <Box sx={{ my: 3, display: 'flex', justifyContent: 'center' }}>
                <form onSubmit={handleSubmit}>
                    <Box sx={{ display: 'flex', justifyContent: 'center', flexWrap: 'wrap' }}>
                        <FormControl sx={{ m: 1 }} variant="outlined">
                            <OutlinedInput
                                id="AdvancedSearchBar-Title"
                                placeholder='Title'
                                name="title"
                                value={formData.title}
                                onChange={handleChange}
                                sx={{ width: { xs: 235, sm: '100%' }, height: { lg: '55px' } }}
                                inputProps={{ sx: { fontSize: { lg: '15.8px' } } }}
                            />
                        </FormControl>

                        <FormControl sx={{ m: 1 }} variant="outlined">
                            <OutlinedInput
                                id="AdvancedSearchBar-Author"
                                placeholder='Author'
                                name="author"
                                value={formData.author}
                                onChange={handleChange}
                                sx={{ width: { xs: 235, sm: '100%' }, height: { lg: '54px' } }}
                                inputProps={{ sx: { fontSize: { lg: '15.8px' } } }}
                            />
                        </FormControl>

                        <FormControl sx={{ m: 1 }} variant="outlined">
                            <OutlinedInput
                                id="AdvancedSearchBar-Keyword"
                                placeholder='Keyword'
                                name="keyword"
                                value={formData.keyword}
                                onChange={handleChange}
                                sx={{ width: { xs: 235, sm: '100%' }, height: { lg: '54px' } }}
                                inputProps={{ sx: { fontSize: { lg: '15.8px' } } }}
                            />
                        </FormControl>

                        <FormControl sx={{ m: 1 }} variant="outlined">
                            <OutlinedInput
                                id="AdvancedSearchBar-ISBN"
                                placeholder='ISBN'
                                name="isbn"
                                value={formData.isbn}
                                onChange={handleChange}
                                sx={{ width: { xs: 235, sm: '100%' }, height: { lg: '54px' } }}
                                inputProps={{ sx: { fontSize: { lg: '15.8px' } } }}
                            />
                        </FormControl>

                        <Box sx={{ mt: { xs: 1, lg: 0 }, width: { xs: '100%', lg: 'auto' }, display: 'flex', justifyContent: 'center' }}>
                            <Button
                                variant="contained"
                                size="medium"
                                sx={{ mt: 1, ml: .4, height: '54px', minWidth: 'unset', width: { xs: '59px', lg: '56px' }, backgroundColor: '#748cab' }}
                                onClick={handleSubmit}
                                type='submit'
                                aria-label="search"
                            >
                                <SearchIcon sx={{ fontSize: { xs: '33px', lg: '31px' } }} />
                            </Button>

                            <Button
                                variant="contained"
                                size="medium"
                                sx={{ mt: 1, ml: 1.5, height: '54px', minWidth: 'unset', width: { xs: '59px', lg: '56px' }, backgroundColor: '#748cab' }}
                                onClick={() => (setOpenAdvancedSearch(false))}
                                aria-label="advanced search"
                            >
                                <i className="fa-solid fa-up-down-left-right fa-2xl"></i>
                            </Button>
                        </Box>
                    </Box>
                </form >
            </Box >
        )
    }

    return (
        <form onSubmit={handleSubmit}>
            <Box sx={{ my: 3, width: '100%', display: 'flex', justifyContent: 'center' }}>
                <FormControl sx={{ m: 1, ml: 2, width: '950px' }} variant="outlined">
                    <OutlinedInput
                        id="SearchBar"
                        placeholder='Search'
                        name="term"
                        value={formData.term}
                        onChange={handleChange}
                        endAdornment={
                            <InputAdornment position="end">
                                <IconButton
                                    size="large"
                                    onClick={handleSubmit}
                                    type='submit'
                                    aria-label="search"
                                >
                                    <SearchIcon />
                                </IconButton>
                            </InputAdornment>
                        }
                    />
                </FormControl>

                <Button
                    variant="contained"
                    sx={{ mt: 1, mr: 2, ml: .5, height: { xs: '55px', lg: '54px' }, minWidth: 'unset', width: '59px', backgroundColor: '#748cab' }}
                    onClick={() => (setOpenAdvancedSearch(true))}
                    aria-label="advanced search"
                >
                    <i className="fa-solid fa-up-down-left-right fa-2xl"></i>
                </Button>
            </Box >
        </form >
    )
};

export default BookSearchForm;
