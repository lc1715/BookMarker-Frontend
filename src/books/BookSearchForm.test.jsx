import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import BookSearchForm from './BookSearchForm';

it('renders the BookSearchForm component', () => {
    render(
        <MemoryRouter>
            <BookSearchForm />
        </MemoryRouter>
    )
})

it('matches with snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <BookSearchForm />
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})