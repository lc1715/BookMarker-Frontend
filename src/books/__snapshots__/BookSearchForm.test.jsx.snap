// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiBox-root css-kvacq2"
    >
      <div
        class="MuiFormControl-root css-18tsngr-MuiFormControl-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id="SearchBar"
            name="term"
            placeholder="Search"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="search"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-oh9jts-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        aria-label="advanced search"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-ge03ab-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <i
          class="fa-solid fa-up-down-left-right fa-2xl"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
